drivers/char/optimsoc-noc.c

```diff
@@ -0,0 +1,436 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * =================================================================
+ *
+ * Linux driver for OpTiMSoC network-on-chip adapter.
+ *
+ * Author(s):
+ *   Pedro H. Penna <pedrohenriquepenna@gmail.com>
+ *   Stefan Wallentowitz <stefan.wallentowitz@tum.de>
+ */
+
+#include <asm/io.h>
+#include <asm/uaccess.h>
+#include <linux/fs.h>
+#include <linux/interrupt.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+
+/**
+ * Maximum number of endpoints in a tile.
+ */
+#define NR_ENDPOINTS 2
+
+/*
+ * Device name.
+ */
+#define OPTIMSOC_NA_NAME "optimsoc-noc"
+
+/*
+ * NoC adapter IRQ number.
+ */
+#define OPTIMSOC_NA_IRQ 4
+
+/*
+ * Base hardware address for NoC adapter.
+ */
+#define OPTIMSOC_NA_BASE_HWADDR 0xe0000000
+
+/**
+ * Base virtual address for NoC adapter
+ */
+static uint32_t OPTIMSOC_NA_BASE_VADDR = 0;
+
+/*
+ * NoC adapter memory size.
+ */
+#define OPTIMSOC_NA_MEM_SIZE (8*1024*1024)
+
+/* 
+ * Address of memory maped registers.
+ */
+#define REG_NUM_CT (OPTIMSOC_NA_BASE_VADDR + 0x28)
+#define BASE       (OPTIMSOC_NA_BASE_VADDR + 0x100000)
+#define REG_NUMEP  BASE
+#define EP_BASE    BASE + 0x2000
+#define EP_OFFSET  0x2000
+#define REG_SEND   0x0
+#define REG_RECV   0x0
+#define REG_ENABLE 0x4
+/**
+ * Retrieves a word at a target address.
+ */
+#define REG32(x) (*((uint32_t *)(x)))
+
+/*
+ * Buffer size (in bytes).
+ */
+#define OPTIMSOC_NA_BUFFER_SIZE 32
+
+/*
+ * Success.
+ */
+#define SUCCESS 0
+
+/*
+ * Failure
+ */
+#define FAILURE -1
+
+/*
+ * Major device number.
+ */
+static int major;
+
+/*
+ * Number of end points.
+ */
+static uint32_t nr_endpoints;
+
+static struct
+{
+	int nopen;                /* Number of processes using this device. */
+	int head;                 /* Buffer head.                           */
+	int tail;                 /* Buffer tail.                           */
+	int data_received;        /* Was any data received?                 */
+	uint32_t *buffer;         /* Input buffer.                          */
+	wait_queue_head_t wqueue; /* Wait queue of processes.               */
+} adapters[NR_ENDPOINTS];
+
+/*
+ * Enables an endpoint.
+ */
+static void enable(unsigned ep)
+{
+	uint32_t *addr;
+
+#if DEBUG
+	printk(KERN_INFO "enable: %x", ep);
+#endif
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_ENABLE);
+
+	*addr = 1;
+}
+
+/*
+ * Sends a word.
+ */
+static void send(unsigned ep, uint32_t word)
+{
+	uint32_t *addr;
+
+#if DEBUG
+	printk(KERN_INFO "send: %x", word);
+#endif
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_SEND);
+
+	*addr = word;
+}
+
+/*
+ * Receives a word.
+ */
+static uint32_t receive(unsigned ep)
+{
+	uint32_t *addr;
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_RECV);
+
+#if DEBUG
+	printk(KERN_INFO "send: %x", *addr);
+#endif
+
+	return (*addr);
+}
+
+
+/*
```

Comment:
Please make this kernel-style documentation comments (https://www.kernel.org/doc/html/latest/doc-guide/kernel-doc.html#writing-kernel-doc-comments)